libraries/eosiolib/contracts/eosio/dispatcher.hpp

```diff
@@ -122,22 +166,36 @@ namespace eosio
 #define EOSIO_DISPATCH_NOTIFY(type, MEMBERS) \
    BOOST_PP_SEQ_FOR_EACH(EOSIO_DISPATCH_NOTIFY_INTERNAL, type, MEMBERS)
 
-#define EOSIO_ACTION_WRAPPER_DECL_1(r, data, action)  \
-   using action = eosio::action_wrapper<BOOST_PP_CAT( \
-       BOOST_PP_STRINGIZE(action), _h), &__contract_class::action, __contract_account>;
+#define EOSIO_ACTION_WRAPPER_DECL_1(r, data, action)                             \
+   using EOSIO_EXTRACT_ACTION_NAME(action) = eosio::action_wrapper<BOOST_PP_CAT( \
+       BOOST_PP_STRINGIZE(EOSIO_EXTRACT_ACTION_NAME(action)), _h),               \
+       &contract_class::EOSIO_EXTRACT_ACTION_NAME(action), contract_account>;
 #define EOSIO_ACTION_WRAPPER_DECL_0(r, data, action)
-
 #define EOSIO_ACTION_WRAPPER_DECL(r, data, action)                                      \
    BOOST_PP_CAT(EOSIO_ACTION_WRAPPER_DECL_, BOOST_PP_COMPL(EOSIO_MATCH_NOTIFY(action))) \
    (r, data, action)
 
+#define EOSIO_REFLECT_ACTION_1(r, data, action)                                          \
+   f(BOOST_PP_CAT(                                                                       \
+       BOOST_PP_STRINGIZE(EOSIO_EXTRACT_ACTION_NAME(action)), _h),                       \
+       eosio::action_type_wrapper<&contract_class::EOSIO_EXTRACT_ACTION_NAME(action)>{}, \
+       EOSIO_GET_RICARDIAN(action) EOSIO_ACTION_ARG_NAME_STRINGS(action));
```

Comment:
I tried giving an action a `ricadrian_contract` and it either vanished silently or was interpreted as an argument name, depending on the order.  In either case, the abigen program compiled and ran with no errors.

Reply:
Odd. I moved it around in token and it generated correctly:

```
   EOSIO_ACTIONS(contract,
                 "eosio.token"_n,
                 action(create, ricardian_contract(create_ricardian), issuer, maximum_supply),
                 action(issue, to, ricardian_contract(issue_ricardian), quantity, memo),
                 action(retire, quantity, memo, ricardian_contract(retire_ricardian)),
                 action(transfer, from, to, ricardian_contract(transfer_ricardian), quantity, memo),
                 action(open, owner, symbol, ram_payer, ricardian_contract(open_ricardian)),
                 action(close, owner, symbol, ricardian_contract(close_ricardian)))
```

Reply:
Look at the spelling of my test more closely.

Reply:
ah. Probably need to add a check for unexpected parens.

Reply:
Added checks for unused action arg names and for unexpected `()` within arg names.